### PR TITLE
cmd/swarm: new localstore database migration

### PR DESCRIPTION
**Pull request by acud**
This PR handles and tests the flow of migrating from the old localstore to the new one.
The update sequence is as follows:
1. User updates the Swarm binary once this gets released
2. Swarm binary checks for old schema name and identifies when the old chunk store format is present and _quits_ the process with exit code 1. User gets alerted. The binary won't start before datastore is migrated
3. User should create a DB export using `swarm db export`
4. User should delete/move old chunk store (`/chunks`)
5. User starts the Swarm binary, new empty chunkstore gets created, user should quit the Swarm binary at this point
6. User should run the DB import command with `--legacy` flag
7. Datastore migration should be now complete

At a minimum the disk space toll is @ `x2` and `x3` at most. We have with @janos decided we do not want to automate this process and we also don't want to add an automatic migration flag to the binary because this will require users to change the way they run Swarm and this change might not be reverted in some cases and can cause trouble in the future (i.e. a User adds a `--migrate` flag to her `runswarm.sh` script and forgets to remove it, cause future migrations to potentially be executed automatically).

RTFM: https://hackmd.io/UAV2uybkQlSerVIX5xUF0g?view

### CODE REVIEW

File: swarm/storage/localstore/schema.go [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+import (
+	"github.com/ethereum/go-ethereum/swarm/log"
+	"github.com/syndtr/goleveldb/leveldb"
+	"github.com/syndtr/goleveldb/leveldb/opt"
+)
+
+// The DB schema we want to use. The actual/current DB schema might differ
+// until migrations are run.
+const CurrentDbSchema = DbSchemaHalloween
```
Comment (zelig) [inline swarm/storage/localstore/schema.go]: should be sanctuary?

File: swarm/swarm.go [hunk 2]
```diff
@@ -148,6 +149,13 @@ func NewSwarm(config *api.Config, mockStore *mock.NodeStore) (self *Swarm, err e
 		resolver = api.NewMultiResolver(opts...)
 		self.dns = resolver
 	}
+	// check that we are not in the old database schema
+	// if so - fail and exit
+	isLegacy := localstore.IsLegacyDatabase(config.ChunkDbPath)
+
+	if isLegacy {
+		return nil, errors.New("Legacy database format detected!")
```
Comment (zelig) [inline swarm/swarm.go]: we need more instructions here

File: cmd/swarm/testdata/datastore_fixture.go [hunk 3]
```diff
@@ -0,0 +1,1390 @@
+package testdata
```
Comment (zelig) [inline cmd/swarm/testdata/datastore_fixture.go:1]: could this be shorter?

File: cmd/swarm/flags.go [hunk 4]
```diff
@@ -182,4 +182,8 @@ var (
 		Usage:  "URL of the Global Store API provider (only for testing)",
 		EnvVar: SwarmGlobalstoreAPI,
 	}
+	SwarmLegacyFlag = cli.BoolFlag{
```
Comment (zelig) [inline cmd/swarm/flags.go:185]: this needs more documemtation

**Review by zelig — CHANGES_REQUESTED** [review]

**Review by zelig — APPROVED** [review]

**Comment by acud** [comment]
@nonsense I have the results for a dry run migration on my local machine. I'm running an `i5` CPU and my SSD is a SATA III SSD, so performance is not so high.
Datastore size was at around 5.5gb.
I added a [resource pooling](https://github.com/ethersphere/go-ethereum/pull/1297/commits/33e1bb277f193c34e68cb93091762ab59385f85d) to throttle the `db.Put`s. I tried 100 and 200 for pool size without a big difference, so I'm pushing 100 at the moment.
tl;dr: export takes 1 minute for 5 gigs, import takes about 15m
```
export
time ./swarm_mig db export /home/growlster/migrationTest/sw/swarm/bzz-9fae04300e275782afafb52f8a988e23d98b85d9/chunks /home/growlster/migrationTest/exphttps://github.com/ethersphere/go-ethereum/pull/1297/commits/33e1bb277f193c34e68cb93091762ab59385f85dort.tar bd83051b5639cadaba5821a57efa131304a2465b522c19d45e5087eece9dc321                                                                                    
INFO [03-28|21:45:56.713] successfully exported 1375147 chunks 
./swarm_mig db export  /home/growlster/migrationTest/export.tar   24.01s user 19.84s system 78% cpu 56.153 total

import (token pool @ 100)
time ./swarm_mig db import /home/growlster/migrationTest/sw/swarm/bzz-9fae04300e275782afafb52f8a988e23d98b85d9/chunks /home/growlster/migrationTest/export.tar bd83051b5639cadaba5821a57efa131304a2465b522c19d45e5087eece9dc321
INFO [03-28|22:08:13.592] successfully imported 1375147 chunks 
./swarm_mig db import  /home/growlster/migrationTest/export.tar   1134.67s user 131.82s system 164% cpu 12:50.79 total

import (token pool @ 200)
time ./swarm_mig db import /home/growlster/migrationTest/sw/swarm/bzz-9fae04300e275782afafb52f8a988e23d98b85d9/chunks /home/growlster/migrationTest/export.tar bd83051b5639cadaba5821a57efa131304a2465b522c19d45e5087eece9dc321
INFO [03-28|22:29:11.723] successfully imported 1375147 chunks 
./swarm_mig db import  /home/growlster/migrationTest/export.tar   1134.91s user 132.98s system 165% cpu 12:47.45 total
```
